third_party/intel/lib/TritonIntelGPUTransforms/OptimizeDotOperands.cpp

```diff
@@ -34,6 +36,94 @@ namespace mlir::triton::gpu::intel {
 
 namespace {
 
+// Represent a def-use chain rooted at 'start' and terminating at tt.trans
+// operation 'end'.
+class Chain {
+  friend raw_ostream &operator<<(raw_ostream &os, const Chain &chain);
+  using Operations = llvm::SmallSetVector<Operation *, 32>;
+
+public:
+  Chain(Operation *start, tt::TransOp end) : start(start), end(end) {
+    assert(start && end && "Expecting valid operations");
+    assert(start != end && "Expecting distinct operations");
+    assert(
+        isTransitivelyUsedBy(start, end) &&
+        "'end' operation should (transitively) use the result of the 'start' "
+        "operation");
+  }
+  bool operator<(const Chain &other) const {
+    return start < other.start || end < other.end;
```

Comment:
if we want to compare address, should it be something like below?
```suggestion
    return start < other.start || &end < &other.end;
```

Reply:
This is fixed